### DISCUSSION TITLE
Mapping references to nested objects

### DISCUSSION THREAD

**Discussion by jasonGFP** [mapstruct/mapstruct#3384] [Q&A] [answered]
I've got an issue in which I've got several classes

```java
class Foo {
    Set<Bar>
    Set<Baz>
}

class Bar{
   //Necesary setting etc
}

class Baz {
   Set<BarSetting>
}

class BarSetting {
    Bar
}
```

The issue i'm dealing with is when  mapping Foo to FooTarget, The set of bar maps over using the bar mapper, the baz set maps over using the baz mapper, but when the baz mapper maps the baz setting, the field referencing the bar object reference the source Bar and not the targetBar that was previously mapped.  Is there a way to use context to map the nested Bar to one of the bars in its only complex object(target) instead of the source complex object.

**Comment by filiphr** [comment]
@jasonGFP I am having troubles understanding what you are trying to achieve. Can you please provide an example mapper and how would you do the mapping without MapStruct?

**Comment by jasonGFP** [comment]
In the above instance there would be a FooMapper


```
            Foo map (Foo sourceFoo){

            Foo targetFoo = new Foo();

            Set<Bar> barSet = new LinkedHashSet<Bar>(Math.max((int) (configs.size() / .75f) + 1, 16));
            for (Bar bar : sourceFoo.getBars) {
                barSet.add(bar);
            }

            targeFoo.setBars(barset);

            //Baz has a set of BarSetting, one field is a reference to sourceFoo.getBazs.get(0)
            Set<Baz> bazSet = new LinkedHashSet<Baz>(Math.max((int) (configs.size() / .75f) + 1, 16));
            for (Baz baz : sourceFoo.getBazs) {
                set.add(baz);
            }

            targetFoo.setBaz(bazSet)
            return targetFoo;
        }

```

The issue I'm running into is if I try to persistTargetFoo, I get a hibernate error because targetFoo.getBazs().get(0).getBar() is a reference to a Bar object in sourceFoo not targetFoo.  Currently I have an @After method that associates all references to the source objects to the the target objects, but it seems unnecessarily manual.

**Reply by filiphr** [reply, marked as answer]
OK I think that I understand. You are trying to create a clone of `Foo`, right? I think that you need some kind of combination of `@DeepClone` and maybe the example that we have for mapping cycles (https://github.com/mapstruct/mapstruct-examples/tree/main/mapstruct-mapping-with-cycles)

**Comment by jasonGFP** [comment]
CycleAvoidingmapper did in fact work, many thanks
